lib/multiplexer/wrappers.go

```diff
@@ -61,6 +69,28 @@ func (c *Conn) Protocol() int {
 	return c.protocol
 }
 
+// Detect detects the connection protocol by peeking into the first few bytes.
+func (c *Conn) Detect() (int, error) {
+	bytes, err := c.reader.Peek(8)
+	if err != nil {
+		return ProtoUnknown, trace.Wrap(err)
+	}
+	proto, err := detectProto(bytes)
+	if err != nil && !trace.IsBadParameter(err) {
+		return ProtoUnknown, trace.Wrap(err)
+	}
+	return proto, nil
+}
+
+// ReadProxyLine reads proxy-line from the connection.
+func (c *Conn) ReadProxyLine() (*ProxyLine, error) {
+	proxyLine, err := ReadProxyLine(c.reader)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Since you have it, might as well store the proxy line to populate RemoteAddr
```suggestion
	}
	c.proxyLine = proxyLine
```